File: src/ts/tests/piece-durations.test.ts
import { expect, test } from 'vitest';
import { Piece, Phrase, Trajectory, Raga } from '../model';
import { Instrument } from '@shared/enums';

// Helper to build a piece with two tracks, second empty
function buildPieceWithEmptyTrack() {
  const raga = new Raga();
  const t1 = new Trajectory({ durTot: 1 });
  const p1 = new Phrase({ trajectories: [t1], raga });
  return new Piece({
    phraseGrid: [[p1], []],
    instrumentation: [Instrument.Sitar, Instrument.Sitar],
    raga,
  });
}

test('durTotFromPhrases creates silent phrase for empty track', () => {
  const piece = buildPieceWithEmptyTrack();
  piece.durTotFromPhrases();
  expect(piece.phraseGrid[1].length).toBe(1);
  const silentTraj = piece.phraseGrid[1][0].trajectories[0];
  expect(silentTraj.id).toBe(12);
  expect(silentTraj.durTot).toBeCloseTo(1);
});

// Helper for NaN trajectory cleanup
function buildPieceWithNaNTraj() {
  const raga = new Raga();
  const piece = new Piece({ raga, instrumentation: [Instrument.Sitar] });
  const good = new Trajectory({ durTot: 1 });
  const bad = new Trajectory({ durTot: NaN });
  const phrase = new Phrase({ trajectories: [good, bad], raga });
  // force NaN durTot before cleaning
  phrase.durTotFromTrajectories();
  piece.phraseGrid[0].push(phrase);
  return { piece, phrase };
}

test('durArrayFromPhrases removes NaN trajectories', () => {
  const { piece, phrase } = buildPieceWithNaNTraj();
  expect(phrase.durTot).toBeNaN();
  piece.durArrayFromPhrases();
  expect(phrase.trajectories.length).toBe(1);
  expect(phrase.durTot).toBeCloseTo(1);
});

File: python/api/tests/piece_durations_test.py
from python.api.classes.piece import Piece, Instrument
from python.api.classes.phrase import Phrase
from python.api.classes.trajectory import Trajectory
from python.api.classes.raga import Raga
import pytest
import math

# Helper to build a piece with two tracks, second empty

def build_piece_with_empty_track():
    raga = Raga()
    t1 = Trajectory(durTot=1)
    p1 = Phrase(trajectories=[t1], raga=raga)
    return Piece(
        phraseGrid=[[p1], []],
        instrumentation=[Instrument.Sitar, Instrument.Sitar],
        raga=raga,
    )


def test_durtotfromphrases_creates_silent_phrase_for_empty_track():
    piece = build_piece_with_empty_track()
    piece.durTotFromPhrases()
    assert len(piece.phraseGrid[1]) == 1
    silent_traj = piece.phraseGrid[1][0].trajectories[0]
    assert silent_traj.id == 12
    assert silent_traj.durTot == pytest.approx(1)

# Helper for NaN trajectory cleanup

def build_piece_with_nan_traj():
    raga = Raga()
    piece = Piece(raga=raga, instrumentation=[Instrument.Sitar])
    good = Trajectory(durTot=1)
    bad = Trajectory(durTot=math.nan)
    phrase = Phrase(trajectories=[good, bad], raga=raga)
    phrase.durTotFromTrajectories()
    piece.phraseGrid[0].append(phrase)
    return piece, phrase


def test_durarrayfromphrases_removes_nan_trajectories():
    piece, phrase = build_piece_with_nan_traj()
    assert math.isnan(phrase.durTot)
    piece.durArrayFromPhrases()
    assert len(phrase.trajectories) == 1
    assert phrase.durTot == pytest.approx(1)
